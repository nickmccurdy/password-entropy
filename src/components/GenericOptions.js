import handleChange from '../handleChange'
import Options from './Options'
import { filter, identity, keys, map, pipe, sum, T } from 'ramda'
import React, { PureComponent } from 'react'
import { Checkbox } from 'react-bootstrap'

export default class GenericOptions extends PureComponent {
  static propTypes = Options.sharedPropTypes
  static shortName = 'Generic'

  static toggles = {
    letters: {
      label: 'Letters',
      example: 'a-z',
      possibleItems: 26
    },
    capitalLetters: {
      label: 'Capital Letters',
      example: 'A-Z',
      possibleItems: 26
    },
    numbers: {
      label: 'Numbers',
      example: '0-9',
      possibleItems: 10
    },
    symbols: {
      label: 'Symbols',
      example: '!@#$%^&*',
      possibleItems: 8
    }
  }

  state = map(T, this.constructor.toggles)

  handleChange = handleChange.bind(this)

  possiblePasswords = () => pipe(filter(identity),
                                 keys,
                                 map(key => this.constructor.toggles[key].possibleItems),
                                 sum)(this.state)

  render () {
    return (
      <Options possiblePasswords={this.possiblePasswords} {...this.props}>
        <h3>Generic</h3>

        {Object.entries(this.constructor.toggles)
               .map(([name, { label, example }]) => (
                 <Checkbox key={name} name={name} checked={this.state[name]} onChange={this.handleChange}>
                   { label } <small>({ example })</small>
                 </Checkbox>
               ))}
      </Options>
    )
  }
}
