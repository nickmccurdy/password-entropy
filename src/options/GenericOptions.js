import { Checkbox } from 'material-ui'
import PropTypes from 'prop-types'
import { always, map, path, T } from 'ramda'
import React from 'react'
import { connect } from 'react-redux'

import { toggleGeneric } from '../actions'

export const toggles = {
  letters: {
    label: 'Letters',
    example: 'a-z',
    possibleItems: 26
  },
  capitalLetters: {
    label: 'Capital Letters',
    example: 'A-Z',
    possibleItems: 26
  },
  numbers: {
    label: 'Numbers',
    example: '0-9',
    possibleItems: 10
  },
  symbols: {
    label: 'Symbols',
    example: '!@#$%^&*',
    possibleItems: 8
  }
}

export const defaultToggles = map(T, toggles)

export function GenericOptions(props) {
  return (
    <div>
      {Object.entries(toggles).map(([name, { label, example }]) => (
        <Checkbox
          key={name}
          name={name}
          label={<div>{ label } <small>({ example })</small></div>}
          checked={props[name]}
          onCheck={props.toggleGeneric}
        />
      ))}
    </div>
  )
}

GenericOptions.propTypes = {
  ...map(always(PropTypes.bool.isRequired), toggles),
  toggleGeneric: PropTypes.func.isRequired
}

export const mapStateToProps = path(['options', 'generic'])

export default connect(mapStateToProps, { toggleGeneric })(GenericOptions)
