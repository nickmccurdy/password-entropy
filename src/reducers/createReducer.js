import { identity } from 'ramda'

export default function createReducer(actionType, initialState, callback = identity) {
  return (state = initialState, { payload, type }) => {
    const matchesAction = type === actionType
    const isValid = typeof payload !== 'number' || payload >= 0

    return matchesAction && isValid ? callback(payload, state) : state
  }
}
