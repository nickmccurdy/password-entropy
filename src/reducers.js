import GenericOptions from './options/GenericOptions'
import { assoc, identity, map, T } from 'ramda'
import { combineReducers } from 'redux'

function createReducer(actionType, initialState, { callback = identity, validator } = {}) {
  return (state = initialState, { payload, type }) => {
    const matchesAction = type === actionType
    const isValid = !validator || validator(payload)

    return matchesAction && isValid ? callback(payload, state) : state
  }
}

export default combineReducers({
  length: createReducer('SET_LENGTH', 6, { validator: n => n >= 1 }),
  options: combineReducers({
    custom: createReducer('SET_CUSTOM', 1, { validator: n => n >= 0 }),
    generic: createReducer ('TOGGLE_GENERIC', map(T, GenericOptions.toggles), {
      callback: ({ checked, name }, state) => assoc(name, checked, state)
    })
  }),
  optionsKey: createReducer('SET_OPTIONS_KEY', 'diceware')
})
