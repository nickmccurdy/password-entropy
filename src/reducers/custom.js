import { path } from 'ramda'

import createReducer from './createReducer'

export default createReducer('SET_CUSTOM', 0)

export function setCustom({ target: { value } }) {
  return { type: 'SET_CUSTOM', payload: parseInt(value, 10) }
}

export const selector = state => ({
  possibleItems: path(['options', 'custom'], state)
})
