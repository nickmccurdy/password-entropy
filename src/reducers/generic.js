import assoc from 'ramda/src/assoc'
import map from 'ramda/src/map'
import path from 'ramda/src/path'
import T from 'ramda/src/T'

import createReducer from './createReducer'

export const toggles = {
  Letters: {
    example: 'a-z',
    possibleItems: 26
  },
  'Capital Letters': {
    example: 'A-Z',
    possibleItems: 26
  },
  Numbers: {
    example: '0-9',
    possibleItems: 10
  },
  Symbols: {
    example: '!@#$%^&*',
    possibleItems: 8
  }
}

export const defaultToggles = map(T, toggles)

export function toggleGenericCallback({ checked, name }, state) {
  return assoc(name, checked, state)
}

export default createReducer(
  'TOGGLE_GENERIC',
  defaultToggles,
  toggleGenericCallback
)

export function toggleGeneric({ target: { name } }, checked) {
  return {
    type: 'TOGGLE_GENERIC',
    payload: {
      checked,
      name
    }
  }
}

export const selector = path(['options', 'Generic'])
