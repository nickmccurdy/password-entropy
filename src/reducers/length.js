import pick from 'ramda/src/pick'

import createReducer from './createReducer'

export default createReducer('SET_LENGTH', 12)

export function setLength({ target: { value } }) {
  return {
    type: 'SET_LENGTH',
    payload: parseInt(value, 10)
  }
}

export const selector = pick(['length'])
