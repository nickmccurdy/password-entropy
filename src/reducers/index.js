import { combineReducers } from 'redux'

import custom from './custom'
import generic from './generic'
import length from './length'
import optionsKey from './optionsKey'

export default combineReducers({
  length,
  options: combineReducers({ custom, generic }),
  optionsKey
})
