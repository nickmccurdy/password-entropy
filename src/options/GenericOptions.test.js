import { shallow } from 'enzyme'
import { GenericOptions, mapStateToProps } from './GenericOptions'
import { map, T } from 'ramda'
import React from 'react'
import reducers from '../reducers'
import { createStore } from 'redux'

const toggles = map(T, GenericOptions.toggles)

it('renders', () => {
  const wrapper = shallow(<GenericOptions toggleGeneric={() => {}} {...toggles}/>)
  expect(wrapper).toMatchSnapshot()
})

test('mapStateToProps', () => {
  const state = createStore(reducers).getState()
  expect(mapStateToProps(state)).toEqual(toggles)
})
