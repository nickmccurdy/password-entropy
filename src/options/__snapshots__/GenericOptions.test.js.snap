// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericOptions 1`] = `
<div>
  <Checkbox
    checked={true}
    disabled={false}
    label={
      <div>
        Letters
         
        <small>
          (
          a-z
          )
        </small>
      </div>
    }
    labelPosition="right"
    name="letters"
    onCheck={[Function]}
  />
  <Checkbox
    checked={true}
    disabled={false}
    label={
      <div>
        Capital Letters
         
        <small>
          (
          A-Z
          )
        </small>
      </div>
    }
    labelPosition="right"
    name="capitalLetters"
    onCheck={[Function]}
  />
  <Checkbox
    checked={true}
    disabled={false}
    label={
      <div>
        Numbers
         
        <small>
          (
          0-9
          )
        </small>
      </div>
    }
    labelPosition="right"
    name="numbers"
    onCheck={[Function]}
  />
  <Checkbox
    checked={true}
    disabled={false}
    label={
      <div>
        Symbols
         
        <small>
          (
          !@#$%^&*
          )
        </small>
      </div>
    }
    labelPosition="right"
    name="symbols"
    onCheck={[Function]}
  />
</div>
`;
